### PR TITLE
fix(ast-spec): remove more invalid properties

### LINKED ISSUE #4759: Remove properties that are used by invalid code
Currently we have the following AST properties declared:

https://github.com/typescript-eslint/typescript-eslint/blob/83294989dad543351a26e95be8d11a91d348679a/packages/ast-spec/src/declaration/TSEnumDeclaration/spec.ts#L13

https://github.com/typescript-eslint/typescript-eslint/blob/83294989dad543351a26e95be8d11a91d348679a/packages/ast-spec/src/declaration/TSModuleDeclaration/spec.ts#L25

https://github.com/typescript-eslint/typescript-eslint/blob/83294989dad543351a26e95be8d11a91d348679a/packages/ast-spec/src/declaration/TSInterfaceDeclaration/spec.ts#L14

https://github.com/typescript-eslint/typescript-eslint/blob/83294989dad543351a26e95be8d11a91d348679a/packages/ast-spec/src/declaration/TSInterfaceDeclaration/spec.ts#L15

These all represent bad code that would cause errors if one attempted to build in TS:
```
public enum Foo {}
// 'public' modifier cannot appear on a module or namespace element.(1044)

abstract namespace Bar {}
// 'abstract' modifier can only appear on a class, method, or property declaration.(1242)

interface Baz implements Bam {}
// Interface declaration cannot have 'implements' clause.(1176)

abstract interface Bam {}
// 'abstract' modifier can only appear on a class, method, or property declaration.
```

TS parses these fine and emits a semantic error, not a syntactic error.
Our parser currently does not error on syntactically bad code, thus we also don't error on it (#1852).

We should remove these properties from the AST as it is non-sensical to include them - nobody should ever be inspecting the code that is produced on them.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
BREAKING CHANGE:
Removes properties from `ast-spec` and therefore `typescript-estree`

## PR Checklist

- [x] Addresses an existing open issue: fixes #4759
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Removes the additional properties listed by @bradzacher in https://github.com/typescript-eslint/typescript-eslint/issues/4759.

### CODE REVIEW

File: packages/typescript-estree/tests/snapshots/typescript/errorRecovery/interface-property-with-default-value.src.ts.shot [hunk 1]
```diff
@@ -8,26 +8,6 @@ Object {
         "body": Array [
           Object {
             "computed": false,
-            "export": undefined,
-            "initializer": Object {
```
Comment (bradzacher) [inline packages/typescript-estree/tests/snapshots/typescript/errorRecovery/interface-property-with-default-value.src.ts.shot:12]: hmm - i wonder if we should throw on this specific case?
[prettier currently supports it](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLWcBOAzAhmOAAgDEIJDgAdKQwgDyUIGcZN0BzQgXlt78IDkAgNzUAvtRAAaEBAAOMVNCbJQuTJggB3AArqEKlLgA2W3AE8VMgEaZ8AazgwAyrgC2cADLo4yPMaY4GzswRxc5fA5kVgBXIJA4N2s4ABMU1M9cKHYY3HY4Ukw3XBhFbOQQXBiYCGkQAAsYN2MAdXrUeCYIgmcDDtQANw7zCrAmKzQoQMwYHTt2Yr8TQJkAKyY6Zw5jOABFGIh4JYD4iMxpiphzOTgmMDYFOrk2WBbUFJh65AAOAAYZZ4QQItOxyCrPW5YAa+GQARwO8Dm8kMlSYAFooHBUqk6pg4PDUHi5nlFkh-CsQIE3Khopg4jImNs4ABBUpsazVOA6LDeTHHCmM7I7faHXxk5bxGC4axvD5fJAAJhkrFwqGMHAAwhA3KSEkwAKx1GKBAAq0sM5PiAziAEkoOlYM57qgFMz7c4rjt+XAxGIgA), but other parsers (babel) throw here.

If we don't throw, I believe prettier will just delete the code silently when it formats.
WDYT?

**Review by bradzacher — APPROVED** [review]
anakin_delete

**Comment by fisker** [comment]
We already [check lot's of errors in our next branch](https://github.com/prettier/prettier/blob/24b24a92d7194d1d3b4411c0ce9d4b62cf10b967/src/language-js/parse/postprocess/typescript.js#L34), so we can test those properties,  it doesn't really matter.

It will be nice to throw errors here, I choose c).
Anyway, it's always good to have a choice, so if not much trouble d) is best for now (we may still need release Prettier 2.x), but after Prettier v3, we definitely don't want hide syntax errors.

**Review by fisker — APPROVED** [review]
Let's kill them ALL!

**Comment by fisker** [comment]
Has `vistorKeys` updated?

File: packages/typescript-estree/tests/snapshots/typescript/errorRecovery/interface-property-with-default-value.src.ts.shot (hunk 1 above)
Comment (fisker) [reply packages/typescript-estree/tests/snapshots/typescript/errorRecovery/interface-property-with-default-value.src.ts.shot:12]: https://github.com/prettier/prettier/pull/14264

**Comment by bradzacher** [comment]
@JoshuaKGoldberg I think we need to be a little stricter here and explicitly error if we encounter this weird bit of syntax.

The issue with just ignoring it is that it'll potentially have downstream effects on ESLint fixers that act on raw source code and could lead to a fixer introducing a syntax error.

**Comment by JoshuaKGoldberg** [comment]
TIL about the (currently) seven `@deprecated` node properties in `typescript.d.ts`:

```ts
interface PropertySignature {
    /** @deprecated A property signature cannot have an initializer */
    readonly initializer?: Expression | undefined;
}
interface PropertyAssignment {
    /** @deprecated A property assignment cannot have a question token */
    readonly questionToken?: QuestionToken | undefined;
    /** @deprecated A property assignment cannot have an exclamation token */
    readonly exclamationToken?: ExclamationToken | undefined;
}
interface ShorthandPropertyAssignment {
    /** @deprecated A shorthand property assignment cannot have modifiers */
    readonly modifiers?: NodeArray<Modifier> | undefined;
    /** @deprecated A shorthand property assignment cannot have a question token */
    readonly questionToken?: QuestionToken | undefined;
    /** @deprecated A shorthand property assignment cannot have an exclamation token */
    readonly exclamationToken?: ExclamationToken | undefined;
}
interface FunctionTypeNode {
    /** @deprecated A function type cannot have modifiers */
    readonly modifiers?: NodeArray<Modifier> | undefined;
}
```

File: packages/typescript-estree/tests/snapshots/typescript/errorRecovery/interface-property-with-default-value.src.ts.shot (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/typescript-estree/tests/snapshots/typescript/errorRecovery/interface-property-with-default-value.src.ts.shot:12]: @bradzacher @fisker I went ahead and added errors for the other `@deprecated` node properties in `typescript.d.ts` (TIL!):

```ts
interface PropertySignature {
    /** @deprecated A property signature cannot have an initializer */
    readonly initializer?: Expression | undefined;
}
interface PropertyAssignment {
    /** @deprecated A property assignment cannot have a question token */
    readonly questionToken?: QuestionToken | undefined;
    /** @deprecated A property assignment cannot have an exclamation token */
    readonly exclamationToken?: ExclamationToken | undefined;
}
interface ShorthandPropertyAssignment {
    /** @deprecated A shorthand property assignment cannot have modifiers */
    readonly modifiers?: NodeArray<Modifier> | undefined;
    /** @deprecated A shorthand property assignment cannot have a question token */
    readonly questionToken?: QuestionToken | undefined;
    /** @deprecated A shorthand property assignment cannot have an exclamation token */
    readonly exclamationToken?: ExclamationToken | undefined;
}
interface FunctionTypeNode {
    /** @deprecated A function type cannot have modifiers */
    readonly modifiers?: NodeArray<Modifier> | undefined;
}
```

https://github.com/typescript-eslint/typescript-eslint/pull/6243#issuecomment-1416905513 / d3c054fc0

Do you think this is reasonable for Prettier too? I'd be interested in sending a PR over if so, I haven't actually contributed to Prettier yet!

Comment (fisker) [reply packages/typescript-estree/tests/snapshots/typescript/errorRecovery/interface-property-with-default-value.src.ts.shot:12]: Great! Saving us from checking those.

File: packages/typescript-estree/src/convert.ts [hunk 2]
```diff
@@ -2466,14 +2503,8 @@ export class Converter {
...
             ? this.convertTypeAnnotation(node.type, node)
             : undefined,
-          initializer:
-            this.convertChild(
-              // eslint-disable-next-line deprecation/deprecation -- TODO breaking change remove this from the AST
-              node.initializer,
-            ) || undefined,
           readonly: hasModifier(SyntaxKind.ReadonlyKeyword, node) || undefined,
           static: hasModifier(SyntaxKind.StaticKeyword, node) || undefined,
-          export: hasModifier(SyntaxKind.ExportKeyword, node) || undefined,
```
Comment (bradzacher) [inline packages/typescript-estree/src/convert.ts:2476]: A few of these removed cases don't have errors assigned to them.
Let's add some - then we can ship this!

**Review by bradzacher — APPROVED** [review]

File: packages/typescript-estree/tests/snapshots/typescript/errorRecovery/interface-property-with-default-value.src.ts.shot (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/typescript-estree/tests/snapshots/typescript/errorRecovery/interface-property-with-default-value.src.ts.shot]: @fisker I poked around in Prettier-land and couldn't figure out if there was anything else for me to change there. I'll defer to you. If you end up filing an issue in Prettier I'd love to tackle it. But it's a bit much for me on my own. 😅

Comment (fisker) [reply packages/typescript-estree/tests/snapshots/typescript/errorRecovery/interface-property-with-default-value.src.ts.shot]: The invalid modifier check is here in this file   https://github.com/prettier/prettier/blob/next/src/language-js/parse/postprocess/typescript.js

Comment (fisker) [reply packages/typescript-estree/tests/snapshots/typescript/errorRecovery/interface-property-with-default-value.src.ts.shot]: And tests are here https://github.com/prettier/prettier/blob/next/tests/format/misc/errors/typescript/modifiers/jsfmt.spec.js

**Pull request merged**